IdentityCore/src/workplacejoin/MSIDWorkPlaceJoinUtilBase.m

```diff
@@ -217,4 +221,161 @@ For asymmetric keys this holds the public key hash which allows digital identity
+        return nil;
+    }
+    
+    NSData *tagData = [kMSIDPrivateKeyIdentifier dataUsingEncoding:NSUTF8StringEncoding];
+    NSString *legacySharedAccessGroup = [NSString stringWithFormat:@"%@.com.microsoft.workplacejoin", teamId];
+    NSDictionary *extraCertAttributes =  @{ (__bridge id)kSecAttrAccessGroup :  legacySharedAccessGroup};
+    
+    // Legacy registrations would have be done using RSA, passing keyType = RSA in query
+    NSMutableDictionary *extraPrivateKeyAttributes = [[NSMutableDictionary alloc] initWithDictionary:@{ (__bridge id)kSecAttrApplicationTag: tagData,
+                                                                                                        (__bridge id)kSecAttrKeyType : (__bridge id)kSecAttrKeyTypeRSA,
+                                                                                                        (__bridge id)kSecAttrAccessGroup : legacySharedAccessGroup}];
+    // For macOS, access group should not be included if kSecUseDataProtectionKeychain = NO as some older versions might throw an error. On macOS, access group should only be specified if kSecUseDataProtectionKeychain  = YES
+#if TARGET_OS_OSX
+    [extraPrivateKeyAttributes removeObjectForKey:(__bridge id)kSecAttrAccessGroup];
+    extraCertAttributes = nil;
+#endif
+    MSID_LOG_WITH_CTX(MSIDLogLevelInfo, context, @"Checking Legacy keychain for registration.");
+    MSIDWPJKeyPairWithCert *legacyKeys = [self findWPJRegistrationInfoWithAdditionalPrivateKeyAttributes:extraPrivateKeyAttributes certAttributes:extraCertAttributes context:context];
+        
+    if (legacyKeys)
+    {
+        legacyKeys.keyChainVersion = MSIDWPJKeychainAccessGroupV1;
+        MSID_LOG_WITH_CTX(MSIDLogLevelInfo, context, @"Returning RSA private device key from legacy registration.");
+        if ([NSString msidIsStringNilOrBlank:tenantId])
+        {
+            // ESTS didn't request a specific tenant, just return default one
+            return legacyKeys;
+        }
+        
+        // Read tenantId for legacy identity
+        NSError *tenantIdError = nil;
+        NSString *registrationTenantId = [MSIDWorkPlaceJoinUtil getWPJStringDataForIdentifier:kMSIDTenantKeyIdentifier context:context error:&tenantIdError];
+        
+        // There's no tenantId on the registration, or it mismatches what server requested, keep looking for a better match. Otherwise, return the identity already.
+        if (!tenantIdError
+            && registrationTenantId
+            && [registrationTenantId isEqualToString:tenantId])
+        {
+            return legacyKeys;
+        }
+    }
+     
+    NSString *tag = nil;
+    __unused MSIDWPJKeyPairWithCert *defaultKeys = nil;
+    NSString *defaultSharedAccessGroup = [NSString stringWithFormat:@"%@.com.microsoft.workplacejoin.v2", teamId];
+    
+    // If tenantId is missing, the caller may have requested for primary registration. Query keychain to get the primary registration tenantId
+    if (tenantId == nil)
```

Comment:
@oldalton  According to Antonio, the tenantId for ECC primary registration is stored separately in keychain and that needs to be queried 1st in case tenantId is not passed as an argument here. However, for PKeyAuth, we get the tenantId in the challenge that is passed in here so primary registration will not be looked up in case of PkeyAuth, Am I correct in that understanding?

Reply:
If you already have a tenantId, then you can just use to query it directly. If you don't have tenantId, then you need this logic.